mkosi/gentoo.py

```diff
@@ -0,0 +1,510 @@
+                {
+                    "sys-kernel/gentoo-kernel-bin",
+                    "sys-firmware/edk2-ovmf",
+                    # "sys-kernel/linux-firmware",
+                }
+            )
+        self.fetch_fix_stage3(root)
+        self.set_profile(args, root)
+        self.set_default_repo(root)
+        self.unmask_arch(root)
+        self.whitelist_licenses(root)
+        self.provide_patches(args, root)
+        self.set_useflags(root)
+        self.gentoo_mkosi_conf(root)
+        self.invoke_emerge(args, root, inside_stage3=False, actions={"--sync"})
+        self.baselayout(args, root)
+        self.update_stage3(args, root)
+
+        if "build-script" in ARG_DEBUG:
+            self.invoke_emerge(args, root, actions={"--info"})
+
+    def fetch_fix_stage3(self, root: str) -> None:
+        """usrmerge tracker bug: https://bugs.gentoo.org/690294"""
+
+        # e.g.: http://distfiles.gentoo.org/releases/amd64/autobuilds/latest-stage3.txt
+        stage3tsf_path_url = urllib.parse.urljoin(
+            self.portage_cfg["GENTOO_MIRRORS"].partition(" ")[0],
+            f"releases/{self.arch}/autobuilds/latest-stage3.txt",
+        )
+        stage3_tar = ""
+        # GENTOO_UPSTREAM: wait for fix upstream: https://bugs.gentoo.org/792081
+        # and more... so we can gladly escape all this hideousness!
+        with urllib.request.urlopen(stage3tsf_path_url) as r:
+            # 20210323T005051Z/stage3-arm64-systemd-20210323T005051Z.tar.xz 196362344
+            # 20210711T170538Z/stage3-amd64-nomultilib-systemd-20210711T170538Z.tar.xz 214470580
+            args_profile = "nomultilib"
+            profilereg = re.compile(f"stage3-{self.arch}-{args_profile}-systemd")
+            lines = list(r)
+            for line in lines:
+                l = line.decode("utf-8")
+                if profilereg.search(l):
+                    stage3_tar, _, _ = l.partition(" ")
+            if not stage3_tar:
+                die("profile names changed upstream?")
+
+        stage3_url_path = urllib.parse.urljoin(
+            self.portage_cfg["GENTOO_MIRRORS"],
+            f"releases/{self.arch}/autobuilds/{stage3_tar}",
+        )
+        stage3_tar_path = os.path.join(self.portage_cfg["DISTDIR"], stage3_tar)
+        if not os.path.isfile(stage3_tar_path):
+            MkosiPrinter.print_step(f"Fetching {stage3_url_path}")
+            os.makedirs(os.path.dirname(stage3_tar_path), exist_ok=True)
+            urllib.request.urlretrieve(stage3_url_path, stage3_tar_path)
+
+        with tarfile.open(stage3_tar_path) as tfd:
+            MkosiPrinter.print_step(f"Extracting {os.path.basename(stage3_tar)}")
+            # GENTOO: usrmerge mess
+            bins = []
+            usr = []
+            usrreg = re.compile("^[.][/](etc|usr|var)")
+            binreg = re.compile("^[.][/](sbin|bin|lib|lib64)")
+            devreg = re.compile("^[.][/]dev")
+            runreg = re.compile("^[.][/]run")
```

Comment:
`re.compile` is generally uncecessary, because `re` caches compiled expressions.